File: project_b_inspection/project_b2/utils/geoUtils.ts

import { LocationPoint } from '../types';

/**
 * Calculates the Haversine distance between two points in kilometers.
 */
export const calculateDistance = (lat1: number, lon1: number, lat2: number, lon2: number): number => {
  const R = 6371; // Radius of the earth in km
  const dLat = deg2rad(lat2 - lat1);
  const dLon = deg2rad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(deg2rad(lat1)) * Math.cos(deg2rad(lat2)) * Math.sin(dLon / 2) * Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
};

const deg2rad = (deg: number): number => deg * (Math.PI / 180);

export const formatSpeed = (speedInMs: number | null): number => {
  if (speedInMs === null) return 0;
  return Math.round(speedInMs * 3.6); // Convert m/s to km/h
};

export const formatDistance = (km: number): string => {
  if (km < 1) return `${(km * 1000).toFixed(0)} m`;
  return `${km.toFixed(2)} km`;
};

export const formatDuration = (ms: number): string => {
  const totalSeconds = Math.floor(ms / 1000);
  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);
  const seconds = totalSeconds % 60;

  return [
    hours > 0 ? `${hours}h` : null,
    minutes > 0 ? `${minutes}m` : null,
    `${seconds}s`
  ].filter(Boolean).join(' ');
};

File: flutter/lib/core/utils/geo_utils.dart
import 'dart:math';

/// RiderMate 2.0 — Geo & Trajectory Math Utilities
/// Ported and enhanced from Project B geoUtils.ts
class GeoUtils {
  /// Calculates Haversine distance between two coordinates in kilometers.
  static double calculateDistance(double lat1, double lon1, double lat2, double lon2) {
    const double r = 6371.0; // Earth radius in km
    final double dLat = _degToRad(lat2 - lat1);
    final double dLon = _degToRad(lon2 - lon1);

    final double a = sin(dLat / 2) * sin(dLat / 2) +
        cos(_degToRad(lat1)) * cos(_degToRad(lat2)) * sin(dLon / 2) * sin(dLon / 2);
    final double c = 2 * atan2(sqrt(a), sqrt(1 - a));
    return r * c;
  }

  static double _degToRad(double deg) => deg * (pi / 180.0);

  /// Converts speed in meters per second to kilometers per hour.
  static int formatSpeed(double? speedInMs) {
    if (speedInMs == null) return 0;
    return (speedInMs * 3.6).round();
  }

  /// Formats distance in kilometers into readable string (e.g. "850 m" or "12.50 km").
  static String formatDistance(double km) {
    if (km < 1.0) {
      return '${(km * 1000).round()} m';
    }
    return '${km.toStringAsFixed(2)} km';
  }

  /// Formats duration in milliseconds into a readable HH:MM:SS string.
  static String formatDuration(int ms) {
    final int totalSeconds = (ms / 1000).floor();
    final int hours = (totalSeconds / 3600).floor();
    final int minutes = ((totalSeconds % 3600) / 60).floor();
    final int seconds = totalSeconds % 60;

    final List<String> parts = [];
    if (hours > 0) parts.add('${hours}h');
    if (minutes > 0) parts.add('${minutes}m');
    parts.add('${seconds}s');

    return parts.join(' ');
  }
}
